src/textual/widget.py

```diff
@@ -243,8 +247,35 @@ def __get__(self, obj: Widget, objtype: type[Widget] | None = None) -> str | Non
         return title.markup
 
 
+_WidgetMetaSub = TypeVar("_WidgetMetaSub", bound="_WidgetMeta")
+
+
+class _WidgetMeta(_MessagePumpMeta):
+    """Metaclass for widgets.
+
+    Used to issue a warning if a widget subclass is created with naming that's
+    incompatible with TCSS/querying.
+    """
+
+    def __new__(
+        mcs: Type[_WidgetMetaSub],
+        name: str,
+        *args: Any,
+        **kwargs: Any,
+    ) -> _WidgetMetaSub:
+        """Hook into widget subclass creation to check the subclass name."""
+        if not name[0].isupper() and not name.startswith("_"):
+            warnings.warn(
+                SyntaxWarning(
+                    f"Widget subclass {name!r} should be capitalised or start with '_'."
+                ),
+                stacklevel=2,
+            )
```

Comment:
Because this is from the std library, suppressing these warnings is “easy”.
Even so, do we add something like a Boolean `textual.widget.SHOW_WIDGET_NAME_WARNING` to make it even easier for users to suppress the warning?

Reply:
(just looking at the code in the comment above; so might be missing context)

Is it intentional that `_foo` seen as valid? (I'm mostly guessing yes, but wanted to be sure).

Reply:
`_foo` is valid and that's ok in the context in which this restriction was introduced.

Ideally, the wording of the warning (and of the comments around the code) should give that away.

I'm open to making it invalid as well if we think it makes sense to be a bit more restrictive with the goal of looking more coherent (given that `foo` raises the warning but `Foo` doesn't).

Reply:
Nah, I think it's fine; just wanted to be 100% sure that the whole package of code was doing what motivated it.